cardano-api/src/Cardano/Api/MetaData.hs

```diff
@@ -0,0 +1,101 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Api.MetaData
+  ( readJSONMetaData
+  , renderMetaDataError
+  ) where
+
+import           Cardano.Prelude hiding (MetaData)
+import           Prelude (String)
+
+import           Control.Exception (IOException)
+import qualified Control.Exception as Exception
+import           Control.Monad.Fail (fail)
+import           Data.Aeson
+import qualified Data.ByteString.Char8 as SC8
+import qualified Data.ByteString.Lazy.Char8 as LC8
+import qualified Data.HashMap.Strict as HMS
+import           Data.Scientific (floatingOrInteger)
+import           Data.Text.Encoding.Base64 (isBase64)
+import qualified Data.Text as Text
+import qualified Data.Vector as Vector
+
+import           Cardano.Api.Error (textShow)
+import           Shelley.Spec.Ledger.MetaData (MetaData(..), MetaDatum(..))
+
+
+data MetaDataError
+  = ConversionErrNullNotAllowed
+  | ConversionErrNumberNotInteger
+  | ConversionErrLongerThan64Bytes
+  | DecodeErrJSON !FilePath !String
+  | ReadFileErr !FilePath !String
+  deriving (Eq, Ord, Show)
+
+
+renderMetaDataError :: MetaDataError -> Text
+renderMetaDataError err =
+  case err of
+    ConversionErrNullNotAllowed -> "JSON Null value is not allowed in MetaData"
+    ConversionErrNumberNotInteger -> "Only integers are allowed in MetaData"
+    ConversionErrLongerThan64Bytes -> "JSON string is longer than 64 bytes"
+    DecodeErrJSON fp decErr ->
+      "Error decoding JSON at: " <> textShow fp <> " Error: " <> textShow decErr
+    ReadFileErr fp rErr ->
+      "Error reading file at: " <> textShow fp <> " Error: " <> textShow rErr
+
+instance FromJSON MetaDatum where
+  parseJSON v =
+    case valueToMetaDatum v of
+      Right mDatum -> return mDatum
+      Left decErr -> fail . show $ renderMetaDataError decErr
+
+instance FromJSON MetaData where
+  parseJSON = withObject "MetaData"  $ \v -> MetaData <$> v .: "mdMap"
```

Comment:
I don't think `MetaData` should be an object containing a key, `mdMap`, that's mapped to a JSON map but, rather, it should just be a JSON map itself.

```suggestion
  parseJSON = withObject "MetaData"  $ \v -> MetaData <$> parseJSON
```